Sources/AsyncHTTPClient/HTTPClient.swift

```diff
@@ -236,67 +299,64 @@ public class HTTPClient {
-                    return channel.pipeline.addHandler(taskHandler)
-                }
+        promise.futureResult.whenComplete { _ in
+            self.stateLock.withLock {
+                self.tasks[task.id] = nil
             }
-
-        if let timeout = self.resolve(timeout: self.configuration.timeout.connect, deadline: deadline) {
-            bootstrap = bootstrap.connectTimeout(timeout)
         }
 
-        let eventLoopChannel: EventLoopFuture<Channel>
-        if request.kind == .unixSocket, let baseURL = request.url.baseURL {
-            eventLoopChannel = bootstrap.connect(unixDomainSocketPath: baseURL.path)
-        } else {
-            let address = self.resolveAddress(request: request, proxy: self.configuration.proxy)
-            eventLoopChannel = bootstrap.connect(host: address.host, port: address.port)
-        }
+        let connection = self.pool.getConnection(for: request, preference: eventLoopPreference, on: taskEL, deadline: deadline)
 
-        eventLoopChannel.map { channel in
-            task.setChannel(channel)
-        }
-        .flatMap { channel in
-            channel.writeAndFlush(request)
-        }
-        .cascadeFailure(to: task.promise)
+        connection.flatMap { connection -> EventLoopFuture<Void> in
+            let channel = connection.channel
+            let addedFuture: EventLoopFuture<Void>
+
+            switch self.configuration.decompression {
+            case .disabled:
+                addedFuture = channel.eventLoop.makeSucceededFuture(())
+            case .enabled(let limit):
+                let decompressHandler = NIOHTTPResponseDecompressor(limit: limit)
+                addedFuture = channel.pipeline.addHandler(decompressHandler, name: "decompressHandler")
+            }
+
+            return addedFuture.flatMap {
+                if let timeout = self.resolve(timeout: self.configuration.timeout.read, deadline: deadline) {
+                    return channel.pipeline.addHandler(IdleStateHandler(readTimeout: timeout), name: "timeoutHandler")
+                } else {
+                    return channel.eventLoop.makeSucceededFuture(())
+                }
+            }.flatMap {
+                let taskHandler = TaskHandler(task: task, delegate: delegate, redirectHandler: redirectHandler, ignoreUncleanSSLShutdown: self.configuration.ignoreUncleanSSLShutdown)
+                return channel.pipeline.addHandler(taskHandler, name: "taskHandler")
+            }.flatMap {
+                task.setConnection(connection)
 
+                let isCancelled = task.lock.withLock {
+                    task.cancelled
+                }
+
+                if !isCancelled {
+                    return channel.writeAndFlush(request)
+                } else {
+                    return channel.eventLoop.makeSucceededFuture(())
+                }
+            }
+        }.whenFailure {
+            promise.fail($0)
```

Comment:
Must review this before merging, might cause connections not to be released if adding an handler was to fail

Reply:
This is related to the latest CI failure, the `testFailingConnectionIsReleased` sometimes causes an `ioOnClosedChannel` here indicating what looks like an incorrect synchronisation

Reply:
Should be fixed